src/ophys_etl/decrosstalk/ica_utils.py

```diff
@@ -0,0 +1,197 @@
+    x = x - m
+    n = np.sqrt(x.shape[0])
+    u, s, v = np.linalg.svd(x, full_matrices=False)
+    w = np.dot(v.T / s, v) * n
+    return np.dot(u, v) * n, w, m
+
+
+def fix_source_assignment(ica_input, ica_output):
+    """
+    Function to rearrange the rows of ica_output so that
+    the first row is the one that most strongly resembles
+    the the first row of ica_input
+
+    Parameters
+    ----------
+    ica_input -- an NxM numpy array where N is the number of
+    channels (2; one for "signal", one for "crosstalk") and
+    M is the number of timesteps. This was the input to the
+    ICA decomposition.
+
+    ica_output -- an NxM numpy array. This was the output of
+    the initial ICA decomposition
+
+    Returns
+    -------
+    An NxM numpy array created by rearranging (if necessary)
+    the rows of ica_output.
+
+    A boolean that is true if the rows of ica_output had to
+    be swapped.
+    """
+    corrs = pearson_ica_in_to_out(ica_input, ica_output)
+    swapped_flag = False
+    if abs(corrs[0]) > abs(corrs[1]):  # sources are not inverted
+        ica_output_corrected = ica_output
+    else:  # sources are inverted, reassign:
+        swapped_flag = True
+        a = copy.deepcopy(ica_output)
+        b = copy.deepcopy(a[0, :])
+        a[0, :] = a[1, :]
+        a[1, :] = b
+        ica_out_swapped = copy.deepcopy(a)
+        ica_output_corrected = copy.deepcopy(ica_out_swapped)
+    return ica_output_corrected, swapped_flag
+
+
+def run_ica(ica_input, iters, seed, verbose=False):
+    """
+    ica_input -- an MxN numpy array; in the context of the decrosstalk
+    problem, N=the number of timesteps; M=2 (first row is signal;
+    second is crosstalk)
+
+    iters -- an int; the number of iterative loops to try to go through
+    to get the off-diagonal elements of the mixing matrix < 0.3
+
+    seed -- an int; the seed of the random number generator that will
+    be fed to sklearn.decompositions.FastICA
+
+    verbose -- if True, also returns a flag indicating whether or not
+    ICA output signals had to be swapped (just used for testing)
+
+    Returns
+    -------
+    ica_output -- and MxN numpy array; ica_output[0,:] is the unmixed signal,
+    ica_output[1,:] is the unmixed crosstalk (in the context of the decrosstalk
+    problem)
+
+    mixing -- the mixing matrix that gets from ica_input to ica_output
+    np.dot(mixing, ica_output) will restore ica_input
+
+    roi_demixed -- a boolean indicating whether or not the iteration to get
+    the off-diagonal elements of the mixing matrix < 0.3 actually worked
+    """
+    # Whiten observations
+    Ow, W, m = whiten_data(ica_input.transpose())
+    alpha = 1
+    beta = 1
+    it = 0
+    roi_demixed = False
+    rng = np.random.RandomState(seed)
+
+    while not roi_demixed and it <= iters:
+        if alpha > 0.3 or beta > 0.3 or alpha < 0 or beta < 0:
+            # Unmixing
+            ica = FastICA(whiten=False, max_iter=10000, random_state=rng)
```

Comment:
sklearn already knows about the bug

https://github.com/scikit-learn/scikit-learn/issues/17162

(though, there doesn't seem to be an attached PR)

My inclination is to leave our fix here because it works. When sklearn addresses that issue, we can compare results to see if they fixed it the way we need them to. I'm not sure I'd be comfortable running our production system off of a fork of sklearn while we wait for them to accept a PR.

Reply:
can we add a `# NOTE` to that effect in the code?